packages/exchange/src/exchange/exchange.py

```diff
@@ -143,24 +144,23 @@ def call_function(self, tool_use: ToolUse) -> ToolResult:
             return ToolResult(tool_use_id=tool_use.id, output=output, is_error=True)
 
         try:
-            if isinstance(tool_use.parameters, dict):
-                output = json.dumps(tool.function(**tool_use.parameters))
-            elif isinstance(tool_use.parameters, list):
-                output = json.dumps(tool.function(*tool_use.parameters))
+            if isinstance(tool_use.parameters, dict) or isinstance(tool_use.parameters, list):
+                function_return_value = tool.function(**tool_use.parameters)
+                if isinstance(function_return_value, dict):
+                    user_decision = function_return_value['user_decision'] if 'user_decision' in function_return_value else None
```

Comment:
i feel this is easier to follow if we flatten out the conditionals

```python
    if isinstance(parameters, dict):
        function_return_value = tool.function(**parameters)
    elif isinstance(parameters, list):
        function_return_value = tool.function(*parameters)
    else:
        raise ValueError(
            f"The provided tool parameters '{parameters}' "
            "could not be interpreted as arguments."
        )

    user_decision = None
    if isinstance(function_return_value, dict):
        user_decision = function_return_value.get("user_decision")
```

Reply:
`call_function` also still has a lot of responsibility, which makes it hard to reason/test. i'd take it one step further by separating out each chunk 

 * `_execute_tool_function(tool, parameters)`
 * `_extract_user_decision(function_return_value)`

so then the final resulting `ToolResult` reads like this

```python
function_return_value = _execute_tool_function(tool, tool_use.parameters)
user_decision = _extract_user_decision(function_return_value)
output = json.dumps(function_return_value)
```